File: ios/Runner/FashionAIHandler.swift
import Foundation
import UIKit
import Accelerate

/**
 * Fashion AI Handler for iOS
 * Uses ONNX Runtime for inference
 * 
 * Provides:
 * 1. CLIP image encoding for embeddings (outfit matching)
 * 2. Clothing classification (category detection)
 * 3. Similarity search for wardrobe items
 */
@available(iOS 13.0, *)
class FashionAIHandler: NSObject {
    
    // MARK: - Constants
    
    static let CLIP_INPUT_SIZE = 224
    static let CLIP_EMBEDDING_DIM = 512
    
    // OpenCLIP ViT-B-32 normalization
    static let CLIP_MEAN: [Float] = [0.48145466, 0.4578275, 0.40821073]
    static let CLIP_STD: [Float] = [0.26862954, 0.26130258, 0.27577711]
    
    // ImageNet normalization for classifier
    static let CLASSIFIER_MEAN: [Float] = [0.485, 0.456, 0.406]
    static let CLASSIFIER_STD: [Float] = [0.229, 0.224, 0.225]
    
    // DeepFashion2 categories
    static let CATEGORIES = [
        "short_sleeve_top", "long_sleeve_top", "short_sleeve_outwear",
        "long_sleeve_outwear", "vest", "sling", "shorts", "trousers",
        "skirt", "short_sleeve_dress", "long_sleeve_dress", "vest_dress", "sling_dress"
    ]
    
    // Category mapping to app categories
    static let CATEGORY_MAPPING: [String: String] = [
        "short_sleeve_top": "Tops",
        "long_sleeve_top": "Tops",
        "short_sleeve_outwear": "Outerwear",
        "long_sleeve_outwear": "Outerwear",
        "vest": "Tops",
        "sling": "Tops",
        "shorts": "Bottoms",
        "trousers": "Bottoms",
        "skirt": "Bottoms",
        "short_sleeve_dress": "Dresses",
        "long_sleeve_dress": "Dresses",
        "vest_dress": "Dresses",
        "sling_dress": "Dresses"
    ]
    
    // Display names for user-friendly output
    static let DISPLAY_NAMES: [String: String] = [
        "short_sleeve_top": "T-Shirt",
        "long_sleeve_top": "Long Sleeve Shirt",
        "short_sleeve_outwear": "Short Jacket",
        "long_sleeve_outwear": "Coat/Jacket",
        "vest": "Vest",
        "sling": "Camisole",
        "shorts": "Shorts",
        "trousers": "Pants",
        "skirt": "Skirt",
        "short_sleeve_dress": "Summer Dress",
        "long_sleeve_dress": "Long Sleeve Dress",
        "vest_dress": "Vest Dress",
        "sling_dress": "Slip Dress"
    ]
    
    // MARK: - Properties
    
    private var isInitialized = false
    private var clipModel: Any?  // ORT Session placeholder
    private var classifierModel: Any?
    
    // Wardrobe index
    private var wardrobeEmbeddings: [[Float]] = []
    private var wardrobePaths: [String] = []
    private var wardrobeMetadata: [[String: Any]] = []
    
    // MARK: - Initialization
    
    override init() {
        super.init()
    }
    
    /// Initialize the AI models
    func initialize() -> Bool {
        // Note: This requires onnxruntime-objc to be added to the project
        // For now, we provide the interface and use CoreML fallback
        
        // Check for ONNX models in bundle
        if let clipPath = Bundle.main.path(forResource: "clip_image_encoder", ofType: "onnx") {
            print("✅ Found CLIP model at: \(clipPath)")
            // TODO: Initialize ONNX Runtime session
        }
        
        if let classifierPath = Bundle.main.path(forResource: "clothing_classifier", ofType: "onnx") {
            print("✅ Found classifier model at: \(classifierPath)")
            // TODO: Initialize ONNX Runtime session
        }
        
        isInitialized = true
        print("✅ FashionAI initialized (models ready for ONNX Runtime)")
        return true
    }
    
    // MARK: - Image Preprocessing
    
    /// Preprocess image for CLIP encoding
    private func preprocessForClip(image: UIImage) -> [Float]? {
        guard let cgImage = image.cgImage else { return nil }
        
        // Resize to 224x224
        let size = CGSize(width: Self.CLIP_INPUT_SIZE, height: Self.CLIP_INPUT_SIZE)
        UIGraphicsBeginImageContextWithOptions(size, false, 1.0)
        image.draw(in: CGRect(origin: .zero, size: size))
        guard let resizedImage = UIGraphicsGetImageFromCurrentImageContext() else {
            UIGraphicsEndImageContext()
            return nil
        }
        UIGraphicsEndImageContext()
        
        guard let resizedCGImage = resizedImage.cgImage else { return nil }
        
        // Get pixel data
        let width = Self.CLIP_INPUT_SIZE
        let height = Self.CLIP_INPUT_SIZE
        let bytesPerPixel = 4
        let bytesPerRow = bytesPerPixel * width
        let bitsPerComponent = 8
        
        var pixelData = [UInt8](repeating: 0, count: width * height * bytesPerPixel)
        let colorSpace = CGColorSpaceCreateDeviceRGB()
        
        guard let context = CGContext(
            data: &pixelData,
            width: width,
            height: height,
            bitsPerComponent: bitsPerComponent,
            bytesPerRow: bytesPerRow,
            space: colorSpace,
            bitmapInfo: CGImageAlphaInfo.premultipliedLast.rawValue
        ) else { return nil }
        
        context.draw(resizedCGImage, in: CGRect(x: 0, y: 0, width: width, height: height))
        
        // Convert to CHW float array with normalization
        var result = [Float](repeating: 0, count: 3 * width * height)
        
        for y in 0..<height {
            for x in 0..<width {
                let offset = (y * width + x) * bytesPerPixel
                let r = Float(pixelData[offset]) / 255.0
                let g = Float(pixelData[offset + 1]) / 255.0
                let b = Float(pixelData[offset + 2]) / 255.0
                
                let idx = y * width + x
                result[idx] = (r - Self.CLIP_MEAN[0]) / Self.CLIP_STD[0]  // R channel
                result[width * height + idx] = (g - Self.CLIP_MEAN[1]) / Self.CLIP_STD[1]  // G
                result[2 * width * height + idx] = (b - Self.CLIP_MEAN[2]) / Self.CLIP_STD[2]  // B
            }
        }
        
        return result
    }
    
    // MARK: - CLIP Encoding
    
    /// Get CLIP embedding for an image
    func getEmbedding(image: UIImage) -> [Float]? {
        guard isInitialized else {
            print("❌ FashionAI not initialized")
            return nil
        }
        
        guard let preprocessed = preprocessForClip(image: image) else {
            print("❌ Failed to preprocess image")
            return nil
        }
        
        // TODO: Run ONNX inference when runtime is available
        // For now, return placeholder or use CoreML handler
        
        // Placeholder: return random embedding (replace with actual inference)
        print("⚠️ ONNX Runtime not available, using placeholder")
        var embedding = [Float](repeating: 0, count: Self.CLIP_EMBEDDING_DIM)
        for i in 0..<Self.CLIP_EMBEDDING_DIM {
            embedding[i] = Float.random(in: -1...1)
        }
        
        // L2 normalize
        let norm = sqrt(embedding.reduce(0) { $0 + $1 * $1 })
        return embedding.map { $0 / norm }
    }
    
    /// Get CLIP embedding from image data
    func getEmbedding(imageData: Data) -> [Float]? {
        guard let image = UIImage(data: imageData) else { return nil }
        return getEmbedding(image: image)
    }
    
    // MARK: - Classification
    
    /// Classify clothing category
    func classify(image: UIImage) -> [String: Float] {
        guard isInitialized else {
            return [:]
        }
        
        // TODO: Run classifier inference
        // For now, use heuristic-based classification from CoreMLHandler
        
        return [:]
    }
    
    // MARK: - Wardrobe Index
    
    /// Load wardrobe embeddings for similarity search
    func loadWardrobeIndex(embeddings: [[Float]], paths: [String], metadata: [[String: Any]]? = nil) {
        wardrobeEmbeddings = embeddings
        wardrobePaths = paths
        wardrobeMetadata = metadata ?? Array(repeating: [:], count: paths.count)
        print("✅ Loaded wardrobe index with \(paths.count) items")
    }
    
    /// Add item to wardrobe index
    func addToWardrobeIndex(image: UIImage, path: String, metadata: [String: Any] = [:]) -> Bool {
        guard let embedding = getEmbedding(image: image) else { return false }
        
        wardrobeEmbeddings.append(embedding)
        wardrobePaths.append(path)
        wardrobeMetadata.append(metadata)
        
        return true
    }
    
    // MARK: - Similarity Search
    
    /// Find similar items in wardrobe
    func findSimilar(queryEmbedding: [Float], topK: Int = 5) -> [(path: String, score: Float, metadata: [String: Any])] {
        guard !wardrobeEmbeddings.isEmpty else { return [] }
        
        // Compute cosine similarities
        var similarities: [(Int, Float)] = []
        
        for (idx, embedding) in wardrobeEmbeddings.enumerated() {
            var dotProduct: Float = 0
            for i in 0..<min(queryEmbedding.count, embedding.count) {
                dotProduct += queryEmbedding[i] * embedding[i]
            }
            similarities.append((idx, dotProduct))
        }
        
        // Sort by similarity (descending) and take top K
        similarities.sort { $0.1 > $1.1 }
        
        return similarities.prefix(topK).map { (idx, score) in
            (path: wardrobePaths[idx], score: score, metadata: wardrobeMetadata[idx])
        }
    }
    
    /// Find similar items to a given image
    func findSimilar(image: UIImage, topK: Int = 5) -> [(path: String, score: Float, metadata: [String: Any])] {
        guard let embedding = getEmbedding(image: image) else { return [] }
        return findSimilar(queryEmbedding: embedding, topK: topK)
    }
    
    // MARK: - Outfit Recommendation
    
    /// Get outfit recommendation based on seed item
    func recommendOutfit(seedImage: UIImage) -> [String: Any] {
        var result: [String: Any] = [:]
        
        // Get seed item classification
        let classification = classify(image: seedImage)
        if let topCategory = classification.max(by: { $0.value < $1.value })?.key {
            result["seed_category"] = topCategory
            result["seed_app_category"] = Self.CATEGORY_MAPPING[topCategory] ?? "Unknown"
            result["display_name"] = Self.DISPLAY_NAMES[topCategory] ?? topCategory
        }
        
        // Get embedding
        if let embedding = getEmbedding(image: seedImage) {
            result["embedding"] = embedding
            
            // Find similar items
            let similar = findSimilar(queryEmbedding: embedding, topK: 10)
            result["similar_items"] = similar.map { item in
                ["path": item.path, "score": item.score, "metadata": item.metadata]
            }
        }
        
        return result
    }
    
    // MARK: - Utility
    
    /// Check if models are loaded
    var isReady: Bool {
        return isInitialized
    }
    
    /// Get wardrobe size
    var wardrobeSize: Int {
        return wardrobePaths.count
    }
}

File: mobile/android/FashionAIHandler.kt
package com.prismstyle.ai

import android.content.Context
import android.graphics.Bitmap
import android.graphics.BitmapFactory
import ai.onnxruntime.*
import java.nio.FloatBuffer
import kotlin.math.sqrt

/**
 * Fashion AI Handler for Android
 * Provides:
 * 1. CLIP image encoding for embeddings (outfit matching)
 * 2. Clothing classification (category detection)
 * 3. Similarity search with FAISS index
 */
class FashionAIHandler(private val context: Context) {
    
    companion object {
        // CLIP model parameters
        private const val CLIP_INPUT_SIZE = 224
        private const val CLIP_EMBEDDING_DIM = 512
        
        // CLIP normalization (OpenCLIP ViT-B-32)
        private val CLIP_MEAN = floatArrayOf(0.48145466f, 0.4578275f, 0.40821073f)
        private val CLIP_STD = floatArrayOf(0.26862954f, 0.26130258f, 0.27577711f)
        
        // Classifier normalization (ImageNet)
        private val CLASSIFIER_MEAN = floatArrayOf(0.485f, 0.456f, 0.406f)
        private val CLASSIFIER_STD = floatArrayOf(0.229f, 0.224f, 0.225f)
        
        // DeepFashion2 categories
        val CATEGORIES = arrayOf(
            "short_sleeve_top", "long_sleeve_top", "short_sleeve_outwear",
            "long_sleeve_outwear", "vest", "sling", "shorts", "trousers",
            "skirt", "short_sleeve_dress", "long_sleeve_dress", "vest_dress", "sling_dress"
        )
        
        // Category to app category mapping
        val CATEGORY_MAPPING = mapOf(
            "short_sleeve_top" to "Tops",
            "long_sleeve_top" to "Tops",
            "short_sleeve_outwear" to "Outerwear",
            "long_sleeve_outwear" to "Outerwear",
            "vest" to "Tops",
            "sling" to "Tops",
            "shorts" to "Bottoms",
            "trousers" to "Bottoms",
            "skirt" to "Bottoms",
            "short_sleeve_dress" to "Dresses",
            "long_sleeve_dress" to "Dresses",
            "vest_dress" to "Dresses",
            "sling_dress" to "Dresses"
        )
    }
    
    private val env: OrtEnvironment = OrtEnvironment.getEnvironment()
    private var clipSession: OrtSession? = null
    private var classifierSession: OrtSession? = null
    private var isInitialized = false
    
    // Wardrobe embeddings for similarity search
    private var wardrobeEmbeddings: Array<FloatArray>? = null
    private var wardrobePaths: Array<String>? = null
    
    /**
     * Initialize the AI models
     */
    fun initialize(): Boolean {
        try {
            // Load CLIP encoder
            val clipBytes = loadModelFromAssets("clip_image_encoder.onnx")
            if (clipBytes != null) {
                val sessionOptions = OrtSession.SessionOptions()
                sessionOptions.setOptimizationLevel(OrtSession.SessionOptions.OptLevel.ALL_OPT)
                clipSession = env.createSession(clipBytes, sessionOptions)
                println("✅ CLIP encoder loaded")
            }
            
            // Load classifier
            val classifierBytes = loadModelFromAssets("clothing_classifier.onnx")
            if (classifierBytes != null) {
                val sessionOptions = OrtSession.SessionOptions()
                sessionOptions.setOptimizationLevel(OrtSession.SessionOptions.OptLevel.ALL_OPT)
                classifierSession = env.createSession(classifierBytes, sessionOptions)
                println("✅ Clothing classifier loaded")
            }
            
            isInitialized = clipSession != null || classifierSession != null
            return isInitialized
            
        } catch (e: Exception) {
            println("❌ Failed to initialize FashionAI: ${e.message}")
            return false
        }
    }
    
    /**
     * Load model from assets folder
     */
    private fun loadModelFromAssets(filename: String): ByteArray? {
        return try {
            context.assets.open("models/$filename").readBytes()
        } catch (e: Exception) {
            println("⚠️ Model not found: $filename")
            null
        }
    }
    
    /**
     * Preprocess image for CLIP encoding
     */
    private fun preprocessForClip(bitmap: Bitmap): FloatArray {
        val resized = Bitmap.createScaledBitmap(bitmap, CLIP_INPUT_SIZE, CLIP_INPUT_SIZE, true)
        val pixels = IntArray(CLIP_INPUT_SIZE * CLIP_INPUT_SIZE)
        resized.getPixels(pixels, 0, CLIP_INPUT_SIZE, 0, 0, CLIP_INPUT_SIZE, CLIP_INPUT_SIZE)
        
        // CHW format with normalization
        val result = FloatArray(3 * CLIP_INPUT_SIZE * CLIP_INPUT_SIZE)
        
        for (i in pixels.indices) {
            val r = ((pixels[i] shr 16) and 0xFF) / 255.0f
            val g = ((pixels[i] shr 8) and 0xFF) / 255.0f
            val b = (pixels[i] and 0xFF) / 255.0f
            
            // RGB channels in CHW format with CLIP normalization
            result[i] = (r - CLIP_MEAN[0]) / CLIP_STD[0]  // R channel
            result[CLIP_INPUT_SIZE * CLIP_INPUT_SIZE + i] = (g - CLIP_MEAN[1]) / CLIP_STD[1]  // G
            result[2 * CLIP_INPUT_SIZE * CLIP_INPUT_SIZE + i] = (b - CLIP_MEAN[2]) / CLIP_STD[2]  // B
        }
        
        return result
    }
    
    /**
     * Get CLIP embedding for an image
     */
    fun getEmbedding(bitmap: Bitmap): FloatArray? {
        if (clipSession == null) {
            println("❌ CLIP session not initialized")
            return null
        }
        
        try {
            val inputData = preprocessForClip(bitmap)
            val dims = longArrayOf(1, 3, CLIP_INPUT_SIZE.toLong(), CLIP_INPUT_SIZE.toLong())
            
            val inputTensor = OnnxTensor.createTensor(env, FloatBuffer.wrap(inputData), dims)
            val results = clipSession!!.run(mapOf("image" to inputTensor))
            
            val output = results[0].value as Array<FloatArray>
            val embedding = output[0]
            
            // L2 normalize
            val norm = sqrt(embedding.map { it * it }.sum())
            return embedding.map { it / norm }.toFloatArray()
            
        } catch (e: Exception) {
            println("❌ Embedding failed: ${e.message}")
            return null
        }
    }
    
    /**
     * Get CLIP embedding from image bytes
     */
    fun getEmbedding(imageData: ByteArray): FloatArray? {
        val bitmap = BitmapFactory.decodeByteArray(imageData, 0, imageData.size)
        return bitmap?.let { getEmbedding(it) }
    }
    
    /**
     * Classify clothing category
     */
    fun classify(bitmap: Bitmap): Map<String, Float> {
        if (classifierSession == null) {
            return emptyMap()
        }
        
        try {
            // Preprocess with ImageNet normalization
            val resized = Bitmap.createScaledBitmap(bitmap, 224, 224, true)
            val pixels = IntArray(224 * 224)
            resized.getPixels(pixels, 0, 224, 0, 0, 224, 224)
            
            val inputData = FloatArray(3 * 224 * 224)
            for (i in pixels.indices) {
                val r = ((pixels[i] shr 16) and 0xFF) / 255.0f
                val g = ((pixels[i] shr 8) and 0xFF) / 255.0f
                val b = (pixels[i] and 0xFF) / 255.0f
                
                inputData[i] = (r - CLASSIFIER_MEAN[0]) / CLASSIFIER_STD[0]
                inputData[224 * 224 + i] = (g - CLASSIFIER_MEAN[1]) / CLASSIFIER_STD[1]
                inputData[2 * 224 * 224 + i] = (b - CLASSIFIER_MEAN[2]) / CLASSIFIER_STD[2]
            }
            
            val dims = longArrayOf(1, 3, 224, 224)
            val inputTensor = OnnxTensor.createTensor(env, FloatBuffer.wrap(inputData), dims)
            val results = classifierSession!!.run(mapOf("input" to inputTensor))
            
            val output = results[0].value as Array<FloatArray>
            val logits = output[0]
            
            // Softmax
            val maxLogit = logits.max()
            val expLogits = logits.map { kotlin.math.exp(it - maxLogit) }
            val sumExp = expLogits.sum()
            val probs = expLogits.map { (it / sumExp).toFloat() }
            
            // Map to categories
            return CATEGORIES.indices.associate { CATEGORIES[it] to probs[it] }
            
        } catch (e: Exception) {
            println("❌ Classification failed: ${e.message}")
            return emptyMap()
        }
    }
    
    /**
     * Load wardrobe embeddings for similarity search
     */
    fun loadWardrobeIndex(embeddings: Array<FloatArray>, paths: Array<String>) {
        wardrobeEmbeddings = embeddings
        wardrobePaths = paths
        println("✅ Loaded wardrobe index with ${paths.size} items")
    }
    
    /**
     * Find similar items in wardrobe
     */
    fun findSimilar(queryEmbedding: FloatArray, topK: Int = 5): List<Pair<String, Float>> {
        val embeddings = wardrobeEmbeddings ?: return emptyList()
        val paths = wardrobePaths ?: return emptyList()
        
        // Compute cosine similarities
        val similarities = embeddings.mapIndexed { idx, embedding ->
            val dotProduct = queryEmbedding.zip(embedding).map { (a, b) -> a * b }.sum()
            Pair(paths[idx], dotProduct)
        }
        
        // Sort by similarity (descending) and take top K
        return similarities.sortedByDescending { it.second }.take(topK)
    }
    
    /**
     * Find similar items to a given image
     */
    fun findSimilarTo(bitmap: Bitmap, topK: Int = 5): List<Pair<String, Float>> {
        val embedding = getEmbedding(bitmap) ?: return emptyList()
        return findSimilar(embedding, topK)
    }
    
    /**
     * Get outfit recommendation based on seed item
     */
    fun recommendOutfit(seedBitmap: Bitmap): Map<String, Any> {
        val result = mutableMapOf<String, Any>()
        
        // Get seed item classification
        val classification = classify(seedBitmap)
        val topCategory = classification.maxByOrNull { it.value }?.key ?: "unknown"
        val appCategory = CATEGORY_MAPPING[topCategory] ?: "Unknown"
        
        result["seed_category"] = topCategory
        result["seed_app_category"] = appCategory
        
        // Get embedding
        val embedding = getEmbedding(seedBitmap)
        if (embedding != null) {
            result["embedding"] = embedding.toList()
            
            // Find similar items (for complementary suggestions)
            val similar = findSimilar(embedding, 10)
            result["similar_items"] = similar.map { mapOf("path" to it.first, "score" to it.second) }
        }
        
        return result
    }
    
    /**
     * Clean up resources
     */
    fun close() {
        clipSession?.close()
        classifierSession?.close()
        env.close()
    }
}
